packages/devtools_app/lib/src/timeline/timeline_model.dart

```diff
@@ -914,72 +956,64 @@ class AsyncTimelineEvent extends TimelineEvent {
-      // and need to be displayed each on a new row.
-      for (AsyncTimelineEvent child in children) {
-        // TODO(kenz): in order to calculate the exact display depth, this needs
-        // to not sum all depths but instead use the [_eventFitsAtDisplayRow]
-        // logic from [FullTimelineEventGroup] to calculate where events fit.
-        // This should work as long as the tree is traversed depth first.
-        displayDepth += child._calculateDisplayDepth();
-      }
-    } else {
-      int maxChildDepth = -1;
-      for (AsyncTimelineEvent child in children) {
-        maxChildDepth = math.max(
-          maxChildDepth,
-          child._calculateDisplayDepth(),
-        );
-      }
-      displayDepth += maxChildDepth;
-    }
-    return _displayDepth = displayDepth;
-  }
+    const currentRow = 0;
+    _displayRows[currentRow].add(this);
 
-  @override
-  bool get hasOverlappingChildren {
-    if (_hasOverlappingChildren != null) return _hasOverlappingChildren;
+    const mainChildRow = currentRow + 1;
     for (int i = 0; i < children.length; i++) {
-      final AsyncTimelineEvent currentChild = children[i];
-      // We do not have to look back because children will be ordered by their
-      // start times.
-      for (int j = i + 1; j < children.length; j++) {
-        final AsyncTimelineEvent sibling = children[j];
-        if (currentChild.isSubtreeOverlapping(sibling)) {
-          return _hasOverlappingChildren = true;
-        }
+      final AsyncTimelineEvent child = children[i];
+      if (i == 0 ||
+          _eventFitsAtDisplayRow(child, mainChildRow, _displayRows.length)) {
+        _mergeChildDisplayRows(mainChildRow, child);
+      } else {
+        // If [child] does not fit on the target row, add it below the current
+        // deepest display row.
+        _mergeChildDisplayRows(displayRows.length, child);
       }
     }
-    return _hasOverlappingChildren = false;
+    return _displayRows;
   }
 
-  bool _hasOverlappingChildren;
-
-  // Warning: this method may be expensive to call for very deep trees.
-  bool isSubtreeOverlapping(TimelineEvent other) {
-    final maxLevelToVerify = math.min(depth, other.depth);
+  bool _eventFitsAtDisplayRow(
+    AsyncTimelineEvent event,
+    int displayRow,
+    int currentLargestRowIndex,
+  ) {
+    final maxLevelToVerify =
+        math.min(event.displayDepth, currentLargestRowIndex - displayRow);
     for (int level = 0; level < maxLevelToVerify; level++) {
-      final lastEventAtLevel = lastChildNodeAtLevel(level);
-      final otherFirstEventAtLevel = other.firstChildNodeAtLevel(level);
-      if (lastEventAtLevel.time.overlaps(otherFirstEventAtLevel.time)) {
-        return true;
+      final lastEventAtLevel = _displayRows[displayRow + level].nullSafeLast();
+      final firstNewEventAtLevel = event.firstChildNodeAtLevel(level);
+      if (lastEventAtLevel != null && firstNewEventAtLevel != null) {
+        final eventsOverlap =
```

Comment:
nit: this code could short circuit a bit earlier.
E.g. return false immediately after each boolean that indicates that the event does not fit.
That would also make it slightly quicker to understand the code as it makes it clear at the top of the method that the if conditions are all separable.

Reply:
good suggestion done.